[AdityaB4] iP

AdityaB4 opened the pull request:
# Chimp 🦧
> "HOO-HOO" - chimp

### Features
1. Adding tasks
2. Finding tasks
3. Deleting tasks

### Steps to use
- Create a `.jar` file
- Use the **jar** file

### Features
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)

More information [here](https://github.com/nus-cs2103-AY2324S2/ip/pull/384)

```java
System.out.println("This is fenced code and I have nothing to put here");
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

AdityaB4 committed:
Add .class files to .gitignore

AdityaB4 committed:
Rename Duke to Chimp and setup greet text

AdityaB4 committed:
Add Echo and Bye

Chimp now repeats what you say

AdityaB4 committed:
Add Storage and Listing

AdityaB4 committed:
Add Task and TaskStatus

TaskStatus is an enum to track marked and unmarked tasks

AdityaB4 committed:
Add mark and unmark functionality, refactor to use switch for commands

AdityaB4 committed:
Add Event subclasses and remove accidental import in Chimp.java

AdityaB4 committed:
Add time strings for Events and Deadlines

AdityaB4 committed:
Add command handling for todo, deadline, event and fix argument parsing bug

AdityaB4 committed:
Uncomment todo case

AdityaB4 committed:
Setup automated text ui testing

AdityaB4 committed:
Add invalid command message and refactor code

AdityaB4 committed:
Add custom Exception classes

AdityaB4 committed:
Extract command handling to method and add error handling

AdityaB4 committed:
Fix todo description bug

todo command previously said todo in the description

AdityaB4 committed:
Add delete command with error handling

AdityaB4 committed:
Add Storage.java and integrate with Chimp.java

AdityaB4 committed:
Change time types to LocalDate

AdityaB4 committed:
Add parsing and error handling for LocalDates

AdityaB4 committed:
Display from, to, by formatted to pattern

AdityaB4 committed:
Update .gitignore to exclude /data

AdityaB4 committed:
Add TaskList.java

AdityaB4 committed:
Fix list function call bug

AdityaB4 committed:
Edit /data folder in .gitignore

AdityaB4 committed:
Add Command package and classes

AdityaB4 committed:
Add Ui, Storage, Parser, TaskList in Core package

AdityaB4 committed:
Create Exception package and add CommandExecuteException

AdityaB4 committed:
Create Task package

AdityaB4 committed:
Integrate changes

AdityaB4 committed:
Change file locations

AdityaB4 committed:
Merge branch 'add-gradle-support'

AdityaB4 committed:
Edit gradle mainClass to support run

AdityaB4 committed:
Setup dummy test

AdityaB4 committed:
Setup tests for Event and Todo

AdityaB4 committed:
Add .jar and /.vscode to gitignore

AdityaB4 committed:
Change constant variable to screaming snake case

AdityaB4 committed:
Rename 'list' -> 'tasks'

AdityaB4 committed:
Use K&R style braces

AdityaB4 committed:
Add FindCommand and corresponding say method

AdityaB4 committed:
Add find command to TaskList

AdityaB4 committed:
Change find return type to Task[]

AdityaB4 committed:
Add JavaDoc comments to TaskList

AdityaB4 committed:
Add JavaDoc comments to Storage

AdityaB4 committed:
Add JavaDoc comments to Parser

AdityaB4 committed:
Add JavaDoc comments to ChimpException

AdityaB4 committed:
Add JavaDoc comments to Chimp

AdityaB4 committed:
Add JavaDoc comments to Ui

AdityaB4 committed:
Add JavaDoc comments to Command

AdityaB4 committed:
Merge branch 'branch-A-CodingStandard'

AdityaB4 committed:
Merge branch 'branch-A-JavaDoc'

AdityaB4 committed:
Fix 'divider' -> 'DIVIDER' after ff merge

AdityaB4 committed:
Fix /data in gitignore

AdityaB4 committed:
Add checkstyle config and gradle task

bennyLCK commented on src/main/java/chimp/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,30 @@
+import chimp.exception.CommandExecuteException;
+
+public class DeadlineCommand extends Command {
+    String text;
```
Perhaps a more meaningful name like taskName would make more semnantic meaning in command classes in general?

bennyLCK commented on src/main/java/chimp/command/DeleteCommand.java:
```diff
@@ -0,0 +1,31 @@
+        this.index = index;
+    }
+
+    @Override
+    public boolean isExit() {
+        return false;
+    }
```
I like how you applied the concept of overriding suitable parent methods instead of what some people would do, accessing the parent's field isExit, and then manually setting and getting that field.

bennyLCK commented on src/main/java/chimp/command/DeleteCommand.java:
```diff
@@ -0,0 +1,31 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws CommandExecuteException{
+        if (index - 1 < 0 || index - 1 >= tasks.size()) {
```
Kuddos to handling the edge cases for invalid integer indexes, but i feel that there can be more error handling for cases where the 'argument' supplied to delete command is say a letter or even empty.

bennyLCK commented on src/main/java/chimp/command/EventCommand.java:
```diff
@@ -0,0 +1,32 @@
+    LocalDate toDate;
+
+    public EventCommand(String text, LocalDate fromDate, LocalDate toDate) {
+        this.text = text;
+        this.fromDate = fromDate;
+        this.toDate = toDate;
```
I like your practice of using fully defined naming references for your instance variables!

bennyLCK commented on src/main/java/chimp/command/FindCommand.java:
```diff
@@ -0,0 +1,27 @@
+import chimp.task.Task;
+
+public class FindCommand extends Command{
+    String arg;
```
Nit: Again, a more meaningful name would allow others to understand your codes better (keyWord for example).

bennyLCK commented on src/main/java/chimp/core/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+import chimp.task.TaskStatus;
+import chimp.task.Todo;
+
+public class TaskList extends ArrayList<Task>{
```
Would it be more appropriate to instead encapsulate an `ArrayList<Task>` as a field in `TaskList` class as a `ArrayList<Task>` object is used to represent the internals of the `TaskList` class?

bennyLCK commented on src/main/java/chimp/command/MarkCommand.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws CommandExecuteException {
+        if (num < 1 || num > tasks.size()) {
+            throw new CommandExecuteException("mark must have number argument");
+        }
```
Perhaps more edge cases can be considered? This also applies to commands of delete, mark/unmark, find etc. that have a number or keyword as additional info used in the commands. Maybe you can consider cases like  letters/word given instead of an index or when it is empty when you are expecting more info.

bennyLCK commented on src/main/java/chimp/core/Parser.java:
```diff
@@ -0,0 +1,96 @@
+                return new FindCommand(arg);
+            default:
+                throw new InvalidCommandException("command \"" + command + "\" is invalid");
+        }
```
Maybe you should try to implement checkstyle or revisit the coding standards webpage? It is not familiar to us, but the `case` statements should not be indented with respect to the `switch` statement based on our shared coding standard. The IDES normally have auto-styling on which indents for us on save / exit so you will have to be careful and prevent your styling changes from being reverted even after you have removed the indents. On Windows, the shortcut would be to highlight the entire block of `case` statements then press shift + tab to shift the block towards the left.

bennyLCK commented on src/main/java/chimp/core/Storage.java:
```diff
@@ -0,0 +1,58 @@
+     * @return true if the command is a save command, false otherwise
+     */
+    public static boolean isSaveCommand(String command) {
+        HashSet<String> saveCommands = new HashSet<>(
```
I like your use of a `HashSet<String>` to keep track of list of allowed commands in your program, thereby reducing the time needed to retrieve the types of a command from instead maybe a list. But perhaps it would be more well suited to be stored and referenced as a static field in the `Command` class instead so that you would not have to instantiate a `saveCommands` object every time you do a predicate check with this method?

bennyLCK commented on src/main/java/chimp/core/Storage.java:
```diff
@@ -0,0 +1,58 @@
+        Path filePath = Path.of("./data/chimp.txt");
+
+        // Create file if it does not exist
+        try {
+            Path directoryPath = filePath.getParent();
+            if (directoryPath != null) {
+                Files.createDirectories(directoryPath);
+            }
+            Files.createFile(filePath);
+        } catch (FileAlreadyExistsException e) {
+            // Ignore this error
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
I like how you separate the actions of creating and writing to a file, but perhaps it would be best to encapsulate the creation of a file as a separate method instead? Especially since we would only be creating a file once at the start.

bennyLCK commented on src/main/java/chimp/core/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    public void add(String task, LocalDate from, LocalDate to) {
+        super.add(new Event(task, TaskStatus.UNMARKED, from, to));
+    }
```
I like your decision of overloading the `add` method and rely on it to return the correct type of task during runtime.

bennyLCK commented on src/main/java/chimp/core/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+        super.add(new Event(task, TaskStatus.UNMARKED, from, to));
+    }
+
+    /**
+     * Returns a string representation of the TaskList.
+     *
+     * @return a string representation of the TaskList
+     */
```
Perhaps the comment here can be omitted entirely since it is trivial? Or the `@return` statement can be omitted here since it is repetitive.

bennyLCK commented on src/main/java/chimp/core/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+     */
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder();
```
I like your use of `StringBuilder` here to increase the efficiency of String concatenation.

bennyLCK commented on src/main/java/chimp/core/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+        return sb.toString();
+    }
+
+    public Task[] find(String keyword) {
+        ArrayList<Task> matches = new ArrayList<>();
+        for (int i = 0; i < this.size(); i++) {
+            if (this.get(i).toString().contains(keyword)) {
+                matches.add(this.get(i));
+            }
+        }
+        return matches.toArray(new Task[0]);
+    }
```
Perhaps you can consider to modify the return type of the method to `ArrayList<Task>` and return `matches` directly? in calling `matches.toArray(new Task[0])`, we would be performing double the work by populating two separate data structures to hold the same data which is wasteful.

bennyLCK commented on src/main/java/chimp/core/Ui.java:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @return A HashMap containing phrases for various actions.
+     */
+    HashMap<String, String> getPhrases() {
+        HashMap<String, String> phrases = new HashMap<>();
+        String greet = " Hello! I'm Chimp\n" +
+                " What can I do for you?\n";
+        String bye = "Bye. Hope to see you again soon!\n";
+        String mark = "Nice! I've marked this task as done: \n";
+        String unmark = "OK, I've marked this task as not done yet: \n";
+        String hoo = "HOO-HOO-HOO-HOO, I don't know what that means\n";
+        String delete = "Noted. I've removed this task:\n";
+        String find = "Here are the matching tasks in your list:\n";
+
+        phrases.put("greet", greet);
+        phrases.put("bye", bye);
+        phrases.put("mark", mark);
+        phrases.put("unmark", unmark);
+        phrases.put("hoo", hoo);
+        phrases.put("delete", delete);
+        phrases.put("find", find);
+
+        return phrases;
+    }
```
Perhaps it would be best to store a `HashMap<String, String> phrases` instance as a static field since you would be referring to it whenever you call the `say` method?

bennyLCK commented on src/main/java/chimp/core/Ui.java:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @param key the key to retrieve the phrase from the hashmap
+     */
+    public void say(String key) {
+        String phrase = getPhrases().get(key);
+        System.out.println(DIVIDER);
+        System.out.println(phrase);
+        System.out.println(DIVIDER);
+    }
```
Perhaps you can consider learning the default `String.format()` method for the Java String class in printing statements with many newline characters, since it can all be grouped up into one print statement?

bennyLCK reviewed (commented):
The overall class design LGTM. Overall, I am very delighted that you used various advanced OOP concepts like overloading, overriding and inheritance to a very large extent. However, there were certain coding standard violations too. Firstly, your classes were missing class header comments which is a requirement, secondly, there were certain methods that kept instantiating new instances of data structures which could have otherwise been avoided, lastly, there were some methods which fulfill more than one purpose, perhaps you can consider extracting different functionalities into its individual methods.

Joshy837 commented on src/main/java/chimp/command/DeleteCommand.java:
```diff
@@ -19,12 +19,13 @@ public boolean isExit() {
     @Override
-    public void execute(TaskList list, Ui ui, Storage storage) throws CommandExecuteException{
-        if (index - 1 < 0 || index - 1 >= list.size())
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws CommandExecuteException{
```
Can add a space after `CommandExecuteException` to follow the code quality standard

Joshy837 commented on src/main/java/chimp/command/UnmarkCommand.java:
```diff
@@ -18,11 +18,12 @@ public boolean isExit() {
     @Override
-    public void execute(TaskList list, Ui ui, Storage storage) throws CommandExecuteException {
-        if (num < 1 || num > list.size())
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws CommandExecuteException {
```
I like how you used the variable name `num` to indicate that the variable used is an integer

Joshy837 commented on src/main/java/chimp/core/Ui.java:
```diff
@@ -4,7 +4,7 @@
 
 public class Ui {
-    private static final String divider = "____________________________________________________________\n";
+    private static final String DIVIDER = "____________________________________________________________\n";
```
I like how you used all caps for `DIVIDER` to explicitly show that it is a constant

Joshy837 reviewed (commented):
Overall good quality code and naming conventions, can make slight improvements to make code even clearer!

AdityaB4 committed:
Add javafx dependencies to build.gradle

AdityaB4 committed:
Extend Application and implement start method

AdityaB4 committed:
Add Launcher

AdityaB4 committed:
Add prompt images

AdityaB4 committed:
Add DialogBox custom control

AdityaB4 committed:
Setup javafx echo dialog

AdityaB4 committed:
Fix build.gradle dependencies

AdityaB4 committed:
Change execute to return String

AdityaB4 committed:
Refactor phrases to static variable

AdityaB4 committed:
Change Ui print methods to String

AdityaB4 committed:
Change DeadlineCommand to return String on execute

AdityaB4 committed:
Change DeleteCommand to return String on execute

AdityaB4 committed:
Change Commands to return String on execute

AdityaB4 committed:
Fix mark/unmark integer parse bug

AdityaB4 committed:
Add javafx layout, formatting and functionality

AdityaB4 committed:
Add interim exit command handling

AdityaB4 committed:
Add assertions to Chimp, Parser, DeleteCommand

The mentioned classes do not use assertions to
document assumptions.

This is improved upon by using assertions
and verifying conditions with low/no overhead.

AdityaB4 committed:
Fix switch case indentation

AdityaB4 committed:
Move guard clause to top of case

AdityaB4 committed:
Adhere to SLAP better

Date extraction and parsing are lower level
functions that are grouped with creation of objects.

Extract out lower level of abstraction into functions
to better adhere to SLAP in some switch cases.

Also clarify happy path by shifting guard clauses to the
top of blocks of code (i.e. functions)

AdityaB4 committed:
Merge pull request #1 from AdityaB4/branch-A-Assertions

Add assertions to Chimp, Parser, DeleteCommand

AdityaB4 committed:
Merge branch 'master' into branch-A-CodeQuality

AdityaB4 committed:
Merge pull request #2 from AdityaB4/branch-A-CodeQuality

Improve code quality principle adherance

AdityaB4 committed:
Add friendlier command shorthands

AdityaB4 committed:
Merge branch 'branch-C-FriendlierSyntax'

AdityaB4 committed:
Replace 'Duke' -> 'Chimp'

AdityaB4 committed:
Update README.md

AdityaB4 committed:
Add blank line before table

AdityaB4 committed:
Add readOutputFromFile and parseLine

AdityaB4 committed:
Update chatbot ui images

Acknowledgement: These images are generated using the DALL-E 2 AI model

AdityaB4 committed:
Add custom flip behaviour to DialogBox control

AdityaB4 committed:
Link flip behaviour in Chimp

AdityaB4 committed:
Fix image encoding

AdityaB4 committed:
Remove ui divider

AdityaB4 committed:
Add line breaks to ui

AdityaB4 committed:
Add Ui screenshot

AdityaB4 committed:
Add file creation function to Storage

AdityaB4 committed:
Link storage functions into ui

AdityaB4 committed:
Fix switch case indentation

AdityaB4 committed:
Fix whitespace in <task>#toString

AdityaB4 committed:
Fix untrimmed string

AdityaB4 committed:
Fix storage parsing logic and regex

AdityaB4 committed:
Add more JavaDoc comments to chimp.task

AdityaB4 committed:
Add more JavaDoc comments to chimp.exception

AdityaB4 committed:
Add more JavaDoc comments to chimp.core

AdityaB4 committed:
Add more JavaDoc comments to chimp.controls

AdityaB4 committed:
Add more JavaDoc comments to chimp.command

AdityaB4 committed:
Add mark and desc tests for Event#toString

AdityaB4 committed:
Add mark and desc tests for Todo#toString

AdityaB4 committed:
Fix EventTest bug

AdityaB4 committed:
Add date format information to docs

AdityaB4 committed:
Link Chimp greeting to ui

AdityaB4 committed:
Fix bye Chimp message

AdityaB4 committed:
Add AI acknowledgement, Separate package statement

AdityaB4 committed:
Delete TodoTest

This commit is part of a filename fix, due to an IDE issue

AdityaB4 committed:
Add TodoTest

This commit is part of a filename fix, due to an IDE issue

AdityaB4 committed:
Change gradle application mainClass to chimp.Launcher

AdityaB4 committed:
Comply with checkstyle (#4)

* Remove trailing whitespaces

* Fix whitespace preceding '{'

* Fix newline at end of file

* Add missing JavaDoc comments

* Fix lexicographical import order

* Add missing access modifiers

* Fix unused param tag

* Fix Chimp imports

* Fix Parser imports

* Fix operator wrap

* Add newline to end of Command

* Fix long lines and imports in Storage
